### PR TITLE
Add contributing guide.

### PR DESCRIPTION
**Pull request by SergioBenitez**
Adds a new `CONTRIBUTING.md` and modifies the README accordingly.

@the10thWiz curious to get your thoughts on this.

[CONTRIBUTING rendered.](https://github.com/rwf2/Rocket/blob/contributing-guide/CONTRIBUTING.md)
[README rendered.](https://github.com/rwf2/Rocket/blob/contributing-guide/README.md)

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+
+First and foremost, **please do not submit a PR that implements a new feature
+without first proposing a design and seeking feedback.** We take the addition of
+new features _very_ seriously because they directly impact usability.
+
+To propose a new feature, create a [new issue] and select "feature request."
+Follow the template, describing your solution in the _ideal solution_ or
+_additional context_ sections as appropriate. Note that certain classes of
+features require strong justification and exceedingly compelling reasons in
+order to be taken into consideration. These include features that:
```
Comment (the10thWiz) [inline CONTRIBUTING.md:90]: I think we should reword this, and provide more detail on what we want from a feature request (most of this is about what not suggest).

Some things I would look for in a feature request:
- Clear issue it's solving (e.g., A would now be possible, or A was very hard, but is now much easier)
- Code examples for what someone using the new feature would look like (and the current work around if relevant)
- Cohesive design
- Limits potential for confusion and foot guns
- As simple as possible (from the perspective of someone using Rocket)

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+  3. **Write basic functionality, pass tests, and submit a PR.**
+
+     Think about edge cases to the problem and ensure you have tests for those
+     edge cases. Once your implementation is functionally complete, submit a PR.
+     Don't spend time writing or changing a bunch of documentation just yet.
+
+  4. **Wait for a review, iterate, and polish.**
+
+     If a review doesn't come in a few days, feel free to ping a maintainer.
+     Someone will review your PR eventually. Once they do, integrate their
```
Comment (the10thWiz) [inline CONTRIBUTING.md:72]:
```suggestion
     Once they review your PR, integrate their
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+  * **Fix** - to fix a functional or doc bug
+    - Example: `Fix 'TcpListener': allow 'udp://' prefix.`
+  * **Improve** - for minor feature or doc improvements
+    - Example: `Improve 'FromParam' derive error messages.`
+  * **Introduce** - for major feature introductions
+    - Example: `Introduce WebSocket support.`
+  * **Add**, **Remove** - for changes
+    - Example: `Add 'Foo::new()' constructor.`
+    - Example: `Remove 'Foo::new()'; add 'Foo::build()'.`
+    - Example: `Update '
```
Comment (the10thWiz) [inline CONTRIBUTING.md:459]:
```suggestion
```

File: README.md [hunk 4]
```diff
@@ -4,7 +4,6 @@
 [![Rocket Homepage](https://img.shields.io/badge/web-rocket.rs-red.svg?style=flat&label=https&colorB=d33847)](https://rocket.rs)
 [![Current Crates.io Version](https://img.shields.io/crates/v/rocket.svg)](https://crates.io/crates/rocket)
 [![Matrix: #rocket:mozilla.org](https://img.shields.io/badge/style-%23rocket:mozilla.org-blue.svg?style=flat&label=[m])](https://chat.mozilla.org/#/room/#rocket:mozilla.org)
-[![IRC: #rocket on irc.libera.chat](https://img.shields.io/badge/style-%23rocket-blue.svg?style=flat&label=Libera.Chat)](https://kiwiirc.com/client/irc.libera.chat/#rocket)
```
Comment (the10thWiz) [inline README.md:7]: Is there an issue with the IRC bridge, or did something else happen?

**Review by the10thWiz — CHANGES_REQUESTED** [review]
Before reviewing, I wrote up a quick skeleton of what I thought should be in this, and the main thing you missed was what tests/examples should look like. Other than that, I think this is a great document.

File: CONTRIBUTING.md (hunk 1 above)
Comment (SergioBenitez) [reply CONTRIBUTING.md:90]: The feature request issue template already has fields for most of these. Do you think it doesn't suffice?

File: README.md (hunk 4 above)
Comment (SergioBenitez) [reply README.md:7]: Yes, the bridge has been off for some time: https://matrix.org/blog/2023/11/28/shutting-down-bridge-to-libera-chat/

File: README.md [hunk 5]
```diff
@@ -44,113 +43,58 @@ Rocket is extensively documented:
...
 cargo run
 ```
 
-You should see `Hello, world!` by visiting `http://localhost:8000`.
-
-## Building and Testing
-
-The `core` directory contains the three core libraries: `lib`, `codegen`, and
-`http` published as `rocket`, `rocket_codegen` and `rocket_http`, respectively.
-The latter two are implementations details and are reexported from `rocket`.
```
[the remark this answers is not in the document]
Comment (SergioBenitez) [reply README.md:74]: I'm not really sure where this fits, and I'm not sure this fits in the contributing guide at all. I feel like this would be best in a completely different type of guide, a development guide of sorts, like [the one for rustc](https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html). But if you can think of a good place for it here, I'm happy to work it in.

**Comment by SergioBenitez** [comment]
Reworked a bit of stuff based on your feedback. Let me know what you think.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+    ├── ...
+    ├── uri_display_type_errors.rs -> ../ui-fail/uri_display_type_errors.rs
+    └── uri_display_type_errors.stderr
+```
+
+If you make changes to codegen, run the UI tests for stable and nightly with
+`test.sh +stable --ui` and `test.sh +nightly --ui`. If there are failures,
+update the outputs with `TRYBUILD=overwrite test.sh +nightly --ui` and
+`TRYBUILD=overwrite test.sh +stable --ui`. Inspect the changes and ensure they
+are reasonable.
```
[the remark this answers is not in the document]
Comment (SergioBenitez) [reply CONTRIBUTING.md]: I think this would also be a good candidate for a "development" guide.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+## Testing
+[Testing]: #testing
+
+All testing happens through [test.sh]. Before submitting a PR, run the script
+and fix any issues. The `default` mode (passing no arguments) will usually
+suffice, but you may also wish to execute another test target. In particular:
+
+  * If you make changes to `contrib`: `test.sh --contrib`
+  * If you make user-facing API changes: `test.sh --examples`
+  * If you add or modify feature flags: `test.sh --core`
```
[the remark this answers is not in the document]
Comment (SergioBenitez) [reply CONTRIBUTING.md]: I wanted to just keep it as generic as possible since you can run it from anywhere.

File: CONTRIBUTING.md (hunk 1 above)
Comment (the10thWiz) [reply CONTRIBUTING.md:88]: It's been a little while since I've opened an issue, so I had to go re-read it. Although I think it covers what we want, I do think there is some value of repeating it here.

Alternatively, we could shorten this section a bit, and simply refer to the issue creation for both what we are, and are not looking for.

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (the10thWiz) [reply README.md:74]: Yeah, a development guide would be a great place to put this. I just wanted to make sure we weren't just deleting important information for new contributors.

**Comment by the10thWiz** [comment]
My main takeaway from the conversation, this time, is that we need a development guide at some point.

Other than that, I just left one comment open. After addressing that (and fixing CI :)), I think we are good to merge.

**Comment by SergioBenitez** [comment]
The CI issues were unrelated and fixed on master. We definitely need a development guide at some point, but it's a big undertaking.

As far as being more specific about new features, what would you suggest, more specifically? Do you think we should improve the issue template and defer to it, add more wording here, or some combination thereof?

**Comment by SergioBenitez** [comment]
Merged.
